### DISCUSSION TITLE
Incorrect type on DataCollection iterator

### DISCUSSION THREAD

**Discussion by iamrgroot** [spatie/laravel-data#202] [Q&A] [answered]
Hi! Thanks for v2! I'm loving the upgrades. However, I'm running into the following phpstan error: "Cannot call method validate() on array." in the following snippet:

```php
/** @var DataCollection<int,Data> */
$collection = Data::collection();

foreach($collection as $item) {
    $item->validate();
}
```

I think this is because the type of [BaseDataCollectable -> getIterator](https://github.com/spatie/laravel-data/blob/main/src/Concerns/BaseDataCollectable.php#L19) is incorrect:

```php
    /**  @return \ArrayIterator<TKey, array> */
    public function getIterator(): ArrayIterator
```

But should be:

```php
    /**  @return \ArrayIterator<TKey, TValue> */
    public function getIterator(): ArrayIterator
```

You can see the difference it makes with with the phpstan playground:
- Current: https://phpstan.org/r/21cf5d37-5693-414c-9b11-918e3dd894d6
- Fixed: https://phpstan.org/r/bfd6646f-8c16-4746-9385-5c89ea5e7fd5

Please correct me if I'm using something wrong!

**Comment by rubenvanassche** [comment, marked as answer]
Looks like you're correct! It will be fixed in the next version!
